### PR TITLE
Make notification center width and height configurable

### PR DESCRIPTION
**Pull request by berbiche**
#33's second commit.

The second commit makes the notification center's width and height configurable.
The option `fit-to-screen` was added, it is incompatible with `notification-center-height` since gtk-layer-shell anchors override the requested height and bind to the edges.

### CODE REVIEW

File: src/controlCenter/controlCenter.vala
Comment (ErikReider) [inline src/controlCenter/controlCenter.vala:120]: Needs to be set in the set_anchor method to be able to update on a config refresh

**Comment by ErikReider** [comment]
Setting the height to 0 seems to shrink it by a lot. Is it possible to set a min height to the ListBox?
image

File: README.md
Comment (ErikReider) [inline README.md:96]: Should also be changed to `control-center-height/width`

**Comment by berbiche** [comment]
I couldn't figure out how to make the height of the window expand automatically to fit all the notifications to display on the screen so I removed the logic for a height of `0`.

I set a minimum width and height of 300 pixels.

File: src/config.json
Comment (ErikReider) [inline src/config.json:12]: Don't forget the `contro-center-heightl`. Other than that it LGTM

**Comment by WTechNinja** [comment]
I may be wrong, but I think `control-center-width` and `notification-window-width` both have a minimum width of 500 from line 41 in src/notification/notification.ui where it says
`<property name="width-request">500</property>`
Could we lower that value to like 300?

**Comment by ErikReider** [comment]
> I may be wrong, but I think `control-center-width` and `notification-window-width` both have a minimum width of 500 from line 41 in src/notification/notification.ui where it says `<property name="width-request">500</property>` Could we lower that value to like 300?

@WTechNinja  Oh yeah, that's true! Hmmm... Setting the notification width from this value would conflict with #42. @berbiche We should probably combine `notification-window-width` with `control-center-width` as `width`. The `control-center-height` should then be `height` and also change the notification windows [max-height](https://github.com/berbiche/SwayNotificationCenter/blob/f11da78c3118a9a1c40c4674a54fa7c2b48c45e3/src/notiWindow/notiWindow.ui#L23)

**Comment by mdjose** [comment]
> I may be wrong, but I think `control-center-width` and `notification-window-width` both have a minimum width of 500 from line 41 in src/notification/notification.ui where it says `<property name="width-request">500</property>` Could we lower that value to like 300?

Even 300px could be a hassle to low resolution screens, instead, it would be nice to allow the user control it via `min-width` and `min-height`.

> @WTechNinja Oh yeah, that's true! Hmmm... Setting the notification width from this value would conflict with #42. @berbiche We should probably combine `notification-window-width` with `control-center-width` as `width`. The `control-center-height` should then be `height` and also change the notification windows [max-height](https://github.com/berbiche/SwayNotificationCenter/blob/f11da78c3118a9a1c40c4674a54fa7c2b48c45e3/src/notiWindow/notiWindow.ui#L23)

I agree with `width` and `height`, but I'm afraid of `max-width` and `max-height` as it imply the notification could autoexpand.

**Comment by berbiche** [comment]
Hi, thanks for the comments and reviews, I'll get around to finish this PR next week.

File: src/notification/notification.ui
Comment (berbiche) [inline src/notification/notification.ui:40]: Removing this property was sufficient for the notification to always expand to the width of the box.

**Comment by berbiche** [comment]
I've updated the PR and took into account the feedback left.

- I did not merge the `notification-window-width` and `control-center-width` because I want to use a different width for each.
- The maximum height of the scrolled window (the list of notifications when the control center is not open) was not modified, it is still limited to 600 pixels.
  I think this feature should be implemented in a different PR with additional logic (e.g. display X notifications to a maximal height of Y pixels for all displayed notifications, only display notifications that can fit in the viewport (don't show half of a notification), etc.)

File: src/controlCenter/controlCenter.vala
Comment (ErikReider) [inline src/controlCenter/controlCenter.vala:158]: Try to make it fit at a 80 char width

**Review by ErikReider — CHANGES_REQUESTED** [review]

**Comment by ErikReider** [comment]
> I think this feature should be implemented in a different PR with additional logic (e.g. display X notifications to a maximal height of Y pixels for all displayed notifications, only display notifications that can fit in the viewport (don't show half of a notification), etc.)

Possibly related to #61. Instead of cutting the notifications off, we could stack them at the bottom

File: src/controlCenter/controlCenter.vala
Comment (ErikReider) [inline src/controlCenter/controlCenter.vala:220]: This only seems to get the primary monitor and not the monitor which the Window is focused on

File: src/controlCenter/controlCenter.vala
Comment (ErikReider) [inline src/controlCenter/controlCenter.vala:213]: Maybe add a short comment above like:
```vala
/** Sets the Window Width and Height and clamps the values to the monitors size */
```

Also this method needs to be called in `on_visibility_change`

File: src/controlCenter/controlCenter.vala
Comment (ErikReider) [inline src/controlCenter/controlCenter.vala:216]: 80 char limit

File: src/configModel/configModel.vala
Comment (ErikReider) [inline src/configModel/configModel.vala:360]: These minimum values should be private

**Review by ErikReider — CHANGES_REQUESTED** [review]

File: src/controlCenter/controlCenter.vala
Comment (ErikReider) [reply src/controlCenter/controlCenter.vala:220]: Maybe we'll need a wlroots vaapi? Hopefully We'll be able to solve it without that but I'm not sure if it'll work with X11 disabled...

Comment (ErikReider) [reply src/controlCenter/controlCenter.vala:220]: Seems like `Gdk.Monitor monitor = this.get_display ().get_monitor_at_window (this.get_window ());` works **BUT** sometime after the map event is called. I'm not sure which event to listen to but `draw` and `check_resize` kinda work but get executed many times...

File: src/controlCenter/controlCenter.vala
Comment (berbiche) [inline src/controlCenter/controlCenter.vala:219]: This is messy and ugly.
Maybe I should refactor the logic to set the anchors to a distinctive function?

File: src/controlCenter/controlCenter.vala
Comment (ErikReider) [inline src/controlCenter/controlCenter.vala:180]: Doesn't seem to work

**Review by ErikReider — CHANGES_REQUESTED** [review]

File: src/controlCenter/controlCenter.vala
Comment (ErikReider) [reply src/controlCenter/controlCenter.vala:219]: Hmmm. Clamping the width and height values seems a little out of reach atm so it would be better to just create a new PR with that enhancement and let this PR focus on setting the width and height

File: src/controlCenter/controlCenter.vala
Comment (berbiche) [reply src/controlCenter/controlCenter.vala:180]: Indeed, this is a leftover.
Will fix!

Comment (berbiche) [reply src/controlCenter/controlCenter.vala:180]: Indeed, this is a leftover.
Will fix!

File: src/controlCenter/controlCenter.vala
Comment (berbiche) [reply src/controlCenter/controlCenter.vala:219]: Done.

File: src/controlCenter/controlCenter.vala
Comment (ErikReider) [inline src/controlCenter/controlCenter.vala:237]: Do we even need this method anymore? Seems like removing this and running `this.set_size_request (configuredWidth, configuredHeight);` in `set_anchor()` reduces the lines by a lot. We can merge this when that's fixed :)

File: src/controlCenter/controlCenter.vala
Comment (ErikReider) [inline src/controlCenter/controlCenter.vala:180]: Removing the `on_map` method results in the X position being flipped. We'd need to switch `anchorLeft` and `anchorRight` to fix this

**Review by ErikReider — CHANGES_REQUESTED** [review]

File: README.md
Comment (ErikReider) [inline README.md:133]: Just added a Man file for the configuration file. Make sure to remove this and add the config options there :)

File: src/controlCenter/controlCenter.vala
Comment (berbiche) [reply src/controlCenter/controlCenter.vala:180]: Indeed, that was a small mistake on my end.

File: src/controlCenter/controlCenter.vala
Comment (berbiche) [reply src/controlCenter/controlCenter.vala:237]: Noted. I didn't realize the code duplication at the time.

File: README.md
Comment (berbiche) [reply README.md]: Done!

**Review by ErikReider — APPROVED** [review]
LGTM! :)

**Pull request merged**
